events/src/test/java/tech/pegasys/artemis/events/EventChannelTest.java

```diff
@@ -183,6 +196,54 @@ public void shouldDeliverEventsAsync() throws Exception {
     inOrder.verifyNoMoreInteractions();
   }
 
+  @Test
+  public void shouldDeliverAsyncEventsOnMultipleThreads() throws Exception {
+    executor =
+        Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("shoudlDeliverAsyncEventsOnMultipleThreads-%d")
+                .build());
+    final EventChannel<WaitOnLatch> channel =
+        EventChannel.createAsync(WaitOnLatch.class, executor, metricsSystem);
+    final WaitOnLatch subscriber =
+        (started, await, completed) -> {
+          started.countDown();
+          try {
+            await.await();
+            completed.countDown();
+          } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+          }
+        };
+    channel.subscribeMultithreaded(subscriber, 2); // Two subscribing threads
+
+    final CountDownLatch started1 = new CountDownLatch(1);
+    final CountDownLatch await1 = new CountDownLatch(1);
+    final CountDownLatch completed1 = new CountDownLatch(1);
+    final CountDownLatch started2 = new CountDownLatch(1);
+    final CountDownLatch await2 = new CountDownLatch(1);
+    final CountDownLatch completed2 = new CountDownLatch(1);
+
+    // Publish two events
+    channel.getPublisher().waitFor(started1, await1, completed1);
+    channel.getPublisher().waitFor(started2, await2, completed2);
+
+    // Both events should start being processed
+    assertThat(started1.await(5, TimeUnit.SECONDS)).isTrue();
```

Comment:
I think spying on started1 and saying `verify(started1).await()` would be more clear here,  because the constants and the time unit makes the reader go look for other constants.

Reply:
You can't use a spy in this case because the test is multithreaded and mockito is generally not thread safe.  I've extracted a private method to make it more readable though.